### DISCUSSION TITLE
Where is the default encryption key

### DISCUSSION THREAD

**Discussion by ygao-armada** [rook/rook#14145] [Q&A]
Recently I turn on OSD volume encryption with uncommenting following line in file deploy/examples/cluster.yaml:
`# encryptedDevice: "true" # the default value for this option is "false"`

Which dmcrypt key is used for the encryption?

I see following secrets:

```
$ kubectl get secret -n rook-ceph
NAME                                TYPE                 DATA   AGE
cluster-peer-token-rook-ceph        kubernetes.io/rook   2      2d17h
rook-ceph-admin-keyring             kubernetes.io/rook   1      2d17h
rook-ceph-config                    kubernetes.io/rook   2      2d17h
rook-ceph-crash-collector-keyring   kubernetes.io/rook   1      2d17h
rook-ceph-dashboard-password        kubernetes.io/rook   1      2d17h
rook-ceph-exporter-keyring          kubernetes.io/rook   1      2d17h
rook-ceph-mgr-a-keyring             kubernetes.io/rook   1      2d17h
rook-ceph-mgr-b-keyring             kubernetes.io/rook   1      2d17h
rook-ceph-mon                       kubernetes.io/rook   4      2d17h
rook-ceph-mons-keyring              kubernetes.io/rook   1      2d17h
rook-csi-cephfs-node                kubernetes.io/rook   2      2d17h
rook-csi-cephfs-provisioner         kubernetes.io/rook   2      2d17h
rook-csi-rbd-node                   kubernetes.io/rook   2      2d17h
rook-csi-rbd-provisioner            kubernetes.io/rook   2      2d17h
```

**Comment by travisn** [comment]
The encryption key is not stored in a K8s secret, but is stored in the Ceph mon store. Are you trying to retrieve the encryption key, or you only want to understand where it is stored?

**Reply by ygao-armada** [reply]
Both to understand and try to retrieve it

**Reply by travisn** [reply]
When `encryptedDevice: true`, Rook passes the `--dmcrypt` option to `ceph-volume` during OSD provisioning. 
@guits Is there a doc that describes how ceph-volume handles the encryption key?

**Reply by guits** [reply]
A specific keyring `client.osd-lockbox.<osd-uuid>` is created for getting the encryption key which is stored in the monitor config store: `ceph -n client.osd-lockbox.<osd-uuid> config-key get dm-crypt/osd/<osd-uuid>/luks`.

This is not documented as the user isn't really supposed to care or play with that.
That being said, I'm not sure what you are trying to achieve but I don't mind adding a section in ceph-volume documentation, that won't hurt, I suppose 🙂 .

**Reply by ygao-armada** [reply]
@guits and @travisn , I try to run following commands, and get permission errors:
```
$ kubectl -n rook-ceph exec -it deploy/rook-ceph-tools -- bash
bash-4.4$ ceph osd ls
0
1
2
3
4
bash-4.4$ ceph osd info 0
osd.0 up   in  weight 1 up_from 42 up_thru 140 down_at 39 last_clean_interval [10,38) [v2:192.168.64.125:6800/1375949130,v1:192.168.64.125:6801/1375949130] [v2:192.168.64.125:6802/1375949130,v1:192.168.64.125:6803/1375949130] exists,up 2b552910-c2fd-45ae-8d38-b8d41986bbf8
bash-4.4$ ceph -n client.osd-lockbox.2b552910-c2fd-45ae-8d38-b8d41986bbf8 config-key get dm-crypt/osd/2b552910-c2fd-45ae-8d38-b8d41986bbf8/luks
[errno 1] RADOS permission error (error connecting to the cluster)
bash-4.4$
```
Am I missing something?

**Reply by guits** [reply]
@ygao-armada I suppose your keyring `client.osd-lockbox.2b552910-c2fd-45ae-8d38-b8d41986bbf8` isn't available within the container you are running that command.

**Reply by travisn** [reply]
@ygao-armada Inside the toolbox of my test minikube cluster, that command works for me:
```
bash-4.4$ ceph osd info 0
osd.0 up   in  weight 1 up_from 77 up_thru 77 down_at 76 last_clean_interval [11,76) [v2:10.244.0.11:6800/1247528763,v1:10.244.0.11:6801/1247528763] [v2:10.244.0.11:6808/1254528763,v1:10.244.0.11:6809/1254528763] exists,up aebc9bb0-0181-4a5b-a437-7bb6732301ab
```

**Reply by ygao-armada** [reply]
@travisn it was the second command not working:
`ceph -n client.osd-lockbox.2b552910-c2fd-45ae-8d38-b8d41986bbf8 config-key get dm-crypt/osd/2b552910-c2fd-45ae-8d38-b8d41986bbf8/luks`

@guits How to get the kering into the pod of command "kubectl -n rook-ceph exec -it deploy/rook-ceph-tools -- bash" ?

**Reply by guits** [reply]
I suppose the question is: has `rook-ceph-tools` enough privileges to export that keyring ?
@travisn ^

**Reply by parth-gr** [reply]
I think we should provide the privileges in the toolbox

**Reply by travisn** [reply]
The toolbox [mounts the admin keyring](https://github.com/rook/rook/blob/master/deploy/examples/toolbox.yaml#L112-L118), so I the toolbox should already have full privs.

**Reply by ygao-armada** [reply]
@travisn how can I leverage the full privs in toolbox to get the encryption key?

**Reply by travisn** [reply]
Does it also fail if you run without the `-n` arg? This will default to the admin keyring.
```
ceph config-key get dm-crypt/osd/2b552910-c2fd-45ae-8d38-b8d41986bbf8/luks
```
 It was probably failing because `-n` was specifying a different user that didn't match the keyring.
